language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -82,79 +115,154 @@ mergePackageMap ps = foldM merge Map.empty ps
 
-dup :: a -> (a, a)
-dup x = (x, x)
+-- Gives the string '@foo' given a directory path like '/path/to/foo'.
+packageSetNameOfPackageSetDir :: FilePath -> String
+packageSetNameOfPackageSetDir = ("@" <>) . takeFileName
 
-infixr 6 <.> -- This is the same fixity as '<>'.
-(<.>) :: T.Text -> T.Text -> T.Text
-(<.>) u v = u <> "." <> v
+-- From the path to a package like '/path/to/daml2ts/d14e08'
+-- calculates '@daml2ts/d14e08' suitable for use as the "name" field
+-- of a 'package.json'.
+packageNameOfPackageDir :: FilePath -> String
+packageNameOfPackageDir packageDir = packageSet <> "/" <> package
+  where
+    packageSet = packageSetNameOfPackageSetDir (takeDirectory packageDir)
+    package = takeFileName packageDir
 
-genModule :: Map.Map PackageId (Maybe PackageName, Package) -> PackageId -> Module -> Maybe T.Text
-genModule pm curPkgId mod
-  | null serDefs = Nothing
+-- Write the files for a single package.
+daml2ts :: Options -> Map.Map PackageId (Maybe PackageName, Package) ->
+    PackageId -> Package -> Maybe PackageName -> String -> String -> IO [String]
+daml2ts Options{..} pm pkgId pkg mbPkgName damlTypesVersion packageVersion = do
+    let packageSetDir = optOutputDir
+          -- The directory into which we generate packages e.g. '/path/to/daml2ts'.
+        packageDir = packageSetDir </> (T.unpack . maybe (unPackageId pkgId) unPackageName) mbPkgName
+          -- The directory into which we write this package e.g. '/path/to/daml2ts/davl-0.0.4'.
+        packageSrcDir = packageDir </> "src"
+          -- Where the source files of this package are written e.g. '/path/to/daml2ts/davl-0.0.4/src'.
+        packageSetName = packageSetNameOfPackageSetDir packageSetDir
+          -- The package set name e.g. '@daml2ts'.
+          -- We use this, for example, when generating import declarations e.g.
+          --   'import * as pkgd14e08_DA_Internal_Template from @daml2ts/d14e08/lib/DA/Internal/Template';'
+    createDirectoryIfMissing True packageSrcDir
+    -- Write .ts files for the package and harvest references to
+    -- foreign packages as we do.
+    T.writeFileUtf8 (packageSrcDir </> "packageId.ts") $ T.unlines
+        ["export default '" <> unPackageId pkgId <> "';"]
+    dependencies <- nubOrd <$> foldM (writeModuleTs packageSrcDir packageSetName) [] (packageModules pkg)
+    -- Now write 'package.json', 'tsconfig.json' and '.eslint.rc.json'
+    -- files into the package dir. The 'package.json' needs the
+    -- dependencies.
+    writeTsConfig packageDir
+    writeEsLintConfig packageDir
+    writePackageJson packageDir damlTypesVersion packageVersion dependencies
+    -- Dependencies have the form '@daml2ts/d14e08'. We want to return
+    -- just the 'd14e08' parts.
+    pure $ mapMaybe (stripPrefix (packageSetName ++ "/")) dependencies
+    where
+      -- Write the .ts file for a single DAML-LF module.
+      writeModuleTs :: FilePath -> String -> [String] -> Module -> IO [String]
```

Comment:
I would recommend to throw a few newtypes at this. E.g. `writeModuleTs :: FilePath -> PackageName -> Module -> IO [Dependency]`.

Reply:
Ok